File: vm/pocketVM.c
// pocketVM.c — Intérprete simples do PocketASM gerado pelo C-ash
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* --------------------- Tipos básicos --------------------- */
typedef struct { int val; char *ccy; } Money;
typedef enum { V_INT, V_STR, V_ACCOUNT, V_MONEY } VKind;

typedef struct {
  VKind k;
  int   own;          // 1 = dono (libera), 0 = não-owning
  union {
    int    i;
    char*  s;
    Money  m;
  } v;
} Value;

typedef struct { Value *data; size_t n, cap; } Stack;
typedef struct { char *name; Value v; } Var;
typedef struct { Var *it; size_t n, cap; } Env;

typedef struct { char *label; int pc; } Label;
typedef struct { char **line; int n; Label *lab; int nlab; } Program;

/* --------------------- Helpers de string/valores --------------------- */
static char* dupstr(const char* s){
  if(!s) return NULL;
  size_t n = strlen(s);
  char* r = malloc(n+1);
  memcpy(r,s,n+1);
  return r;
}

static Value v_copy(const Value* a){
  Value r = *a;
  if (a->k==V_STR || a->k==V_ACCOUNT){
    r.v.s = a->v.s ? dupstr(a->v.s) : NULL;
    r.own = 1;
  } else if (a->k==V_MONEY){
    r.v.m.ccy = a->v.m.ccy ? dupstr(a->v.m.ccy) : NULL;
    r.own = 1;
  } else {
    r.own = 0;
  }
  return r;
}

static void v_free(Value* v){
  if(!v) return;
  if (v->own){
    if (v->k==V_STR || v->k==V_ACCOUNT){
      free(v->v.s);
    } else if (v->k==V_MONEY){
      free(v->v.m.ccy);
    }
  }
  v->own = 0;
}

/* --------------------- Stack --------------------- */
static void push(Stack *S, Value v){
  if(S->n==S->cap){
    S->cap = S->cap ? S->cap*2 : 16;
    S->data = realloc(S->data, S->cap*sizeof(Value));
  }
  S->data[S->n++] = v;
}
static Value pop(Stack *S){
  if(!S->n){ fprintf(stderr,"stack underflow\n"); exit(1); }
  return S->data[--S->n];
}
static Value top(Stack *S){
  if(!S->n){ fprintf(stderr,"stack empty\n"); exit(1); }
  return S->data[S->n-1];
}

/* --------------------- Ambiente --------------------- */
static Var* env_find(Env* E, const char* name){
  for(size_t i=0;i<E->n;i++)
    if(strcmp(E->it[i].name,name)==0) return &E->it[i];
  return NULL;
}

static Value env_get(Env* E, const char* name){
  Var* v = env_find(E,name);
  if(!v){
    fprintf(stderr,"[env] undefined var '%s'\n",name);
    Value z={.k=V_INT,.v.i=0};
    return z;
  }
  return v_copy(&v->v);
}

static void env_set(Env* E, const char* name, Value v){
  Var* ex = env_find(E,name);
  if(ex){
    v_free(&ex->v);
    ex->v = v;
    return;
  }
  if(E->n==E->cap){
    E->cap = E->cap ? E->cap*2 : 8;
    E->it = realloc(E->it, E->cap*sizeof(Var));
  }
  E->it[E->n].name = dupstr(name);
  E->it[E->n].v = v;
  E->n++;
}

static void env_clear(Env* E){
  for(size_t i=0;i<E->n;i++){
    free(E->it[i].name);
    v_free(&E->it[i].v);
  }
  free(E->it);
  E->it=NULL; E->n=E->cap=0;
}

/* --------------------- Snapshots --------------------- */
typedef struct Snap {
  Env env;
  struct Snap* prev;
} Snap;

static Snap* tx_push(Snap* top, Env* E){
  Snap* s = calloc(1,sizeof(Snap));
  s->prev = top;
  for(size_t i=0;i<E->n;i++){
    Var* v = &E->it[i];
    Value nv = v_copy(&v->v);
    env_set(&s->env, v->name, nv);
  }
  return s;
}

static void tx_pop_commit(Snap** top){
  Snap* s=*top; if(!s) return;
  env_clear(&s->env);
  *top=s->prev;
  free(s);
}

static void tx_pop_rollback(Snap** top, Env* E){
  Snap* s=*top; if(!s) return;
  env_clear(E);
  *E = s->env;
  *top = s->prev;
  free(s);
}

/* --------------------- Loader ASM --------------------- */
static char* rstrip(char* p){ int n=strlen(p); while(n>0&&(p[n-1]=='\n'||p[n-1]=='\r')) p[--n]=0; return p; }
static int is_blank(const char* s){ while(*s){ if(!isspace((unsigned char)*s)) return 0; s++; } return 1; }
static void trim_leading(char** s){ while(**s && isspace((unsigned char)**s)) (*s)++; }

static Program load_asm(const char* path){
  Program P={0};
  FILE* f=fopen(path,"rb");
  if(!f){ perror("open .asm"); exit(1); }
  char buf[4096];
  while(fgets(buf,sizeof(buf),f)){
    rstrip(buf);
    char* line = dupstr(buf);
    P.line = realloc(P.line, (P.n+1)*sizeof(char*));
    P.line[P.n++] = line;
  }
  fclose(f);
  for(int i=0;i<P.n;i++){
    char* s=P.line[i]; trim_leading(&s);
    if(s[0]=='.' && strchr(s,':')){
      char* colon = strchr(s,':');
      size_t L = colon-s;
      char* lab = malloc(L+1);
      memcpy(lab,s,L); lab[L]=0;
      P.lab = realloc(P.lab,(P.nlab+1)*sizeof(Label));
      P.lab[P.nlab++] = (Label){ .label=lab, .pc=i };
    }
  }
  return P;
}

static int find_label(Program* P, const char* lbl){
  for(int i=0;i<P->nlab;i++)
    if(strcmp(P->lab[i].label,lbl)==0) return P->lab[i].pc;
  fprintf(stderr,"label não encontrada: %s\n",lbl);
  exit(1);
}

/* --------------------- Execução --------------------- */
static void print_val(Value v){
  switch(v.k){
    case V_INT:    printf("%d", v.v.i); break;
    case V_STR:    printf("%s", v.v.s?v.v.s:""); break;
    case V_ACCOUNT:printf("<acct:%s>", v.v.s?v.v.s:"?"); break;
    case V_MONEY:  printf("%d$ %s", v.v.m.val, v.v.m.ccy?v.v.m.ccy:"???"); break;
  }
}

int main(int argc,char**argv){
  if(argc!=2){ fprintf(stderr,"uso: %s file.asm\n",argv[0]); return 2; }

  Program P=load_asm(argv[1]);
  Stack S={0};
  Env E={0};
  Snap* TX=NULL;

  for(int pc=0; pc<P.n; pc++){
    char* raw=P.line[pc];
    char* s=raw; trim_leading(&s);
    if(is_blank(s)||s[0]==';'||(s[0]=='.'&&strchr(s,':'))) continue;

    char op[64]; sscanf(s,"%63s",op);
    char* rest=s+strlen(op); while(*rest==' '||*rest=='\t') rest++;

    /* ---------------- Instruções ---------------- */
    if(strcmp(op,"PUSH_I")==0){
      int v; sscanf(rest,"%d",&v);
      Value x={.k=V_INT,.v.i=v};
      push(&S,x);
    }
    else if(strcmp(op,"PUSH_STR")==0){
      char* q1=strchr(rest,'"'); char* q2=q1?strrchr(q1,'"'):NULL;
      size_t n=q2&&q2>q1?(size_t)(q2-q1-1):0;
      char* s2=malloc(n+1); memcpy(s2,q1+1,n); s2[n]=0;
      Value x={.k=V_STR,.own=1,.v.s=s2};
      push(&S,x);
    }
    else if(strcmp(op,"PUSH_MONEY")==0){
      int v=0; char ccy[8]="";
      if(sscanf(rest,"%d , %7s",&v,ccy)!=2 && sscanf(rest,"%d %7s",&v,ccy)!=2){
        fprintf(stderr,"PUSH_MONEY malformado: %s\n",rest); exit(1);
      }
      Value x={.k=V_MONEY,.own=1};
      x.v.m.val=v; x.v.m.ccy=dupstr(ccy);
      push(&S,x);
    }
    else if(strcmp(op,"LOAD")==0){
      char name[128]; sscanf(rest,"%127s",name);
      push(&S,env_get(&E,name));
    }
    else if(strcmp(op,"STORE")==0){
      char name[128]; sscanf(rest,"%127s",name);
      Value v=pop(&S);
      env_set(&E,name,v_copy(&v));
      v_free(&v);
    }
    else if(strcmp(op,"ADD")==0||strcmp(op,"SUB")==0||strcmp(op,"MUL")==0||
            strcmp(op,"DIV")==0||strcmp(op,"MOD")==0){
      Value b=pop(&S),a=pop(&S);
      int ai=a.v.i, bi=b.v.i, r=0;
      if(strcmp(op,"ADD")==0)r=ai+bi;
      else if(strcmp(op,"SUB")==0)r=ai-bi;
      else if(strcmp(op,"MUL")==0)r=ai*bi;
      else if(strcmp(op,"DIV")==0)r=bi?ai/bi:0;
      else if(strcmp(op,"MOD")==0)r=bi?ai%bi:0;
      v_free(&a); v_free(&b);
      Value x={.k=V_INT,.v.i=r}; push(&S,x);
    }
    else if(strcmp(op,"CMP_EQ")==0||strcmp(op,"CMP_NE")==0||
            strcmp(op,"CMP_LT")==0||strcmp(op,"CMP_GT")==0||
            strcmp(op,"CMP_LE")==0||strcmp(op,"CMP_GE")==0){
      Value b=pop(&S),a=pop(&S);
      int ai=a.v.i, bi=b.v.i, r=0;
      if(strcmp(op,"CMP_LT")==0)r=ai<bi;
      else if(strcmp(op,"CMP_GT")==0)r=ai>bi;
      else if(strcmp(op,"CMP_LE")==0)r=ai<=bi;
      else if(strcmp(op,"CMP_GE")==0)r=ai>=bi;
      else if(strcmp(op,"CMP_EQ")==0)r=ai==bi;
      else if(strcmp(op,"CMP_NE")==0)r=ai!=bi;
      v_free(&a); v_free(&b);
      Value x={.k=V_INT,.v.i=r}; push(&S,x);
    }
    else if(strcmp(op,"DUP")==0){
      Value a=top(&S);
      push(&S,v_copy(&a));
    }
    else if(strcmp(op,"POP")==0){
      Value a=pop(&S); v_free(&a);
    }
    else if(strcmp(op,"JZ")==0||strcmp(op,"JNZ")==0){
      char lab[64]; sscanf(rest,"%63s",lab);
      Value a=pop(&S);
      int cond=(a.k==V_INT)?a.v.i:0;
      v_free(&a);
      if((strcmp(op,"JZ")==0 && !cond)||(strcmp(op,"JNZ")==0 && cond))
        pc=find_label(&P,lab)-1;
    }
    else if(strcmp(op,"JMP")==0){
      char lab[64]; sscanf(rest,"%63s",lab);
      pc=find_label(&P,lab)-1;
    }
    else if(strcmp(op,"CALL")==0){
      char name[128]; sscanf(rest,"%127s",name);

      if(strcmp(name,"display")==0||strcmp(name,"print")==0){
        Value a=pop(&S);
        print_val(a); printf("\n");
        v_free(&a);
      }
      else if(strncmp(name,"_open_",6)==0){
        const char* bank=name+6;
        char handle[256]; snprintf(handle,sizeof(handle),"%s:acct",bank);
        Value acct={.k=V_ACCOUNT,.own=1,.v.s=dupstr(handle)};
        push(&S,acct);
      }
      else if(strcmp(name,"enforce")==0){
        Value b=pop(&S),a=pop(&S);
        v_free(&a); v_free(&b);
      }
      else if(strcmp(name,"xfer")==0){
        Value amount=pop(&S),to=pop(&S),from=pop(&S);
        printf("[VM] xfer "); print_val(amount);
        printf(" from "); print_val(from);
        printf(" to "); print_val(to); printf("\n");
        v_free(&from); v_free(&to); v_free(&amount);
      }
      else if(strcmp(name,"now")==0||strcmp(name,"region")==0||
              strcmp(name,"net_ok")==0||strcmp(name,"fx_rate")==0){
        Value x={.k=V_INT,.v.i=1}; push(&S,x);
      }
      else printf("[VM] CALL %s (stub)\n",name);
    }
    else if(strcmp(op,"TX_BEGIN")==0){
      TX=tx_push(TX,&E);
      printf("[VM] TX_BEGIN\n");
    }
    else if(strcmp(op,"TX_COMMIT")==0){
      printf("[VM] TX_COMMIT\n");
      tx_pop_commit(&TX);
    }
    else if(strcmp(op,"TX_ROLLBACK")==0){
      printf("[VM] TX_ROLLBACK\n");
      tx_pop_rollback(&TX,&E);
    }
    else if(strcmp(op,"RET")==0){
      break;
    }
  }

  /* limpeza final */
  for(int i=0;i<P.n;i++) free(P.line[i]);
  for(int i=0;i<P.nlab;i++) free(P.lab[i].label);
  free(P.line); free(P.lab);
  env_clear(&E);
  free(S.data);
  while(TX) tx_pop_commit(&TX);
  return 0;
}

File: vm/pocketVM.py
# pocketVM.py — PocketASM VM (funções, validações e currency checks)
import sys, re

K_INT, K_STR, K_ACCOUNT, K_MONEY = "int", "str", "account", "money"

def v_int(i): return {"k": K_INT, "v": i}
def v_str(s): return {"k": K_STR, "v": s}
def v_acct(s): return {"k": K_ACCOUNT, "v": s}
def v_money(v,c): return {"k": K_MONEY, "v": {"val": v, "ccy": c}}
def v_copy(v): return {"k": v["k"], "v": v["v"].copy() if isinstance(v["v"], dict) else v["v"]}

class Stack:
    def __init__(self): self.data = []
    def push(self,v): self.data.append(v)
    def pop(self):
        if not self.data: raise IndexError("stack underflow")
        return self.data.pop()
    def top(self):
        if not self.data: raise IndexError("stack empty")
        return self.data[-1]
    def __len__(self): return len(self.data)

class Env:
    def __init__(self): self.map = {}
    def get(self,name): return v_copy(self.map.get(name, v_int(0)))
    def set(self,name,v): self.map[name]=v_copy(v)
    def clone(self):
        e=Env()
        e.map={k:v_copy(v) for k,v in self.map.items()}
        return e

class Snap:
    def __init__(self,env): self.env=env.clone(); self.prev=None

class Program:
    def __init__(self):
        self.lines=[]
        self.labels={}
        self.bank_limits={}
        self.bank_currency={}
        self.declared_banks=set()
        self.func_labels={}

def trim(s): return s.strip()

def load_asm(path):
    P=Program()
    with open(path, encoding="utf-8") as f:
        P.lines=[l.rstrip("\r\n") for l in f]

    current_bank = None
    for i,raw in enumerate(P.lines):
        s=trim(raw)
        if s.startswith(".") and ":" in s:
            lab=s.split(":",1)[0]
            P.labels[lab]=i

        # bancos e políticas
        if s.startswith("; BANK"):
            m=re.match(r";\s*BANK\s+(\w+)\s+currency\s*=\s*([A-Z]{3})",s)
            if m:
                name,ccy=m.groups()
                current_bank = name
                P.declared_banks.add(name)
                P.bank_currency[name]=ccy
                continue

        if s.startswith("; POLICY limit"):
            m=re.match(r";\s*POLICY\s+limit\s+\S+\s+(\d+)\$ ?([A-Z]{3})",s)
            if m and current_bank:
                P.bank_limits[current_bank]=(int(m.group(1)),m.group(2))
                continue

        # funções
        if s.startswith("; FN "):
            m=re.match(r";\s*FN\s+(\w+)",s)
            if m:
                fn=m.group(1)
                for j in range(i, len(P.lines)):
                    if P.lines[j].strip().startswith(f".FN_{fn}:"):
                        P.func_labels[fn]=j
                        break
    return P

def print_val(v):
    k=v["k"]
    if k==K_INT: print(v["v"], end="")
    elif k==K_STR: print(v["v"], end="")
    elif k==K_ACCOUNT: print(f"<acct:{v['v']}>", end="")
    elif k==K_MONEY: print(f"{v['v']['val']}${v['v']['ccy']}", end="")

def parse_str(rest):
    q1, q2 = rest.find('"'), rest.rfind('"')
    return rest[q1+1:q2] if q1>=0 and q2>q1 else ""

def parse_money(rest):
    m=re.match(r"\s*(\d+)\s*[,$ ]\s*([A-Za-z]{3})",rest)
    if not m: raise RuntimeError(f"[VM] PUSH_MONEY malformado: {rest}")
    return int(m.group(1)), m.group(2)

# ... (cabeçalho igual)

class Frame:
    def __init__(self, ret_pc, prev_env, args):
        self.ret_pc = ret_pc
        self.prev_env = prev_env
        self.args = list(args)

def run(path):
    P = load_asm(path)
    S = Stack()
    E = Env()
    TX = None
    call_frames = [] 

    pc = 0
    n = len(P.lines)

    skip_mode = False
    for i, line in enumerate(P.lines):
        s = trim(line)
        if not s or s.startswith(";"):
            continue

        if s.startswith(".FN_"):
            skip_mode = True
            continue

        if skip_mode:
            if s.startswith("RET"):
                skip_mode = False
            continue

        if not s.startswith(".") and not s.startswith(";"):
            pc = i
            break

    while pc < n:
        raw = P.lines[pc]
        s = trim(raw)
        pc += 1

        if not s or s.startswith(";"): 
            continue
        if s.startswith(".") and ":" in s:
            continue

        parts = s.split(None, 1)
        op = parts[0]
        rest = parts[1].split(";")[0].strip() if len(parts) > 1 else ""

        try:
            if op == "PUSH_I":
                S.push(v_int(int(rest) if rest else 0))

            elif op == "PUSH_STR":
                S.push(v_str(parse_str(rest)))

            elif op == "PUSH_MONEY":
                val, ccy = parse_money(rest)
                S.push(v_money(val, ccy))

            elif op == "LOAD":
                S.push(E.get(rest))

            elif op == "STORE":
                E.set(rest, S.pop())

            elif op in ("ADD","SUB","MUL","DIV","MOD"):
                b, a = S.pop(), S.pop()
                if a["k"]==K_MONEY or b["k"]==K_MONEY:
                    if a["k"]!=b["k"] or a["v"]["ccy"]!=b["v"]["ccy"]:
                        raise RuntimeError("[VM] operação entre moedas diferentes.")
                    av, bv = a["v"]["val"], b["v"]["val"]
                    if op=="ADD":   S.push(v_money(av+bv, a["v"]["ccy"]))
                    elif op=="SUB": S.push(v_money(av-bv, a["v"]["ccy"]))
                    elif op=="MUL": S.push(v_money(av*bv, a["v"]["ccy"]))
                    elif op=="DIV": S.push(v_money(av//bv if bv else 0, a["v"]["ccy"]))
                    else:           S.push(v_money(av%bv if bv else 0, a["v"]["ccy"]))
                else:
                    av, bv = a["v"], b["v"]
                    if op=="ADD":   S.push(v_int(av+bv))
                    elif op=="SUB": S.push(v_int(av-bv))
                    elif op=="MUL": S.push(v_int(av*bv))
                    elif op=="DIV": S.push(v_int(av//bv if bv else 0))
                    else:           S.push(v_int(av%bv if bv else 0))

            elif op in ("JZ","JNZ"):
                lab = rest
                cond = S.pop()
                c = cond["v"] if cond["k"] == K_INT else 0
                if (op=="JZ" and not c) or (op=="JNZ" and c):
                    pc = P.labels.get(lab, pc)

            elif op == "JMP":
                pc = P.labels.get(rest, pc)

            elif op == "TX_BEGIN":
                snap = Snap(E); snap.prev = TX; TX = snap
                print("[VM] TX_BEGIN")

            elif op == "TX_COMMIT":
                print("[VM] TX_COMMIT")
                TX = TX.prev if TX else None

            elif op == "TX_ROLLBACK":
                print("[VM] TX_ROLLBACK")
                if TX:
                    E = TX.env.clone()
                    TX = TX.prev

            elif op == "CALLFN":
                m = re.match(r"(\w+)\s+(\d+)", rest)
                if not m:
                    raise RuntimeError(f"[VM] CALLFN malformado: {rest}")
                name, argc_s = m.groups()
                argc = int(argc_s)

                tmp = []
                for _ in range(argc):
                    tmp.append(S.pop())
                args = list(reversed(tmp))

                if name not in P.func_labels:
                    print(f"[VM] CALL {name} (stub)")
                    continue

                fr = Frame(ret_pc=pc, prev_env=E, args=args)
                call_frames.append(fr)
                E = Env()
                pc = P.func_labels[name] + 1
                continue

            elif op == "PUSH_ARG":
                idx = int(rest) if rest else 0
                if not call_frames:
                    raise RuntimeError("PUSH_ARG fora de função")
                fr = call_frames[-1]
                if idx < 0 or idx >= len(fr.args):
                    S.push(v_int(0))
                else:
                    S.push(v_copy(fr.args[idx]))

            elif op == "CALL":
                name = rest

                if name in ("display","print"):
                    a = S.pop()
                    print_val(a); print()
                    continue

                if name.startswith("_open_"):
                    bank = name[6:]
                    if bank not in P.declared_banks:
                        print(f"[VM] ERRO: banco '{bank}' não declarado.")
                        return 1
                    S.push(v_acct(f"{bank}:acct"))
                    continue

                if name == "enforce":
                    _b, _a = S.pop(), S.pop()
                    continue

                if name == "xfer":
                    amount, to, from_ = S.pop(), S.pop(), S.pop()
                    print(f"[VM] xfer {amount['v']['val']}${amount['v']['ccy']} from {from_['v']} to {to['v']}")
                    if from_["k"] == K_ACCOUNT:
                        bank = from_["v"].split(":",1)[0]
                        if bank not in P.bank_currency:
                            print(f"[VM] ERRO: banco '{bank}' sem currency declarada.")
                            return 1
                        bank_ccy = P.bank_currency[bank]
                        if amount["v"]["ccy"] != bank_ccy:
                            print(f"[VM] ERRO: não é possível transferir {amount['v']['val']}${amount['v']['ccy']} de uma conta {bank_ccy} sem conversão explícita.")
                            return 1
                        lim = P.bank_limits.get(bank)
                        if lim and amount["v"]["val"] > lim[0]:
                            print(f"[VM] LIMITE EXCEDIDO: {amount['v']['val']}${amount['v']['ccy']} > {lim[0]}${lim[1]}")
                            print("[VM] TX_ROLLBACK")
                            print("Transação revertida com sucesso.")
                            return 0
                    continue

                if name in ("now","region","net_ok","fx_rate"):
                    S.push(v_int(1))
                    continue

                print(f"[VM] CALL {name} (stub)")

            elif op == "RET":
                if call_frames:
                    fr = call_frames.pop()
                    pc = fr.ret_pc
                    E = fr.prev_env
                else:
                    break

        except Exception as e:
            sys.stderr.write(f"[VM] erro em '{raw}': {e}\n")
            return 1

    return 0


if __name__=="__main__":
    if len(sys.argv)!=2:
        print(f"uso: {sys.argv[0]} file.asm")
        sys.exit(2)
    sys.exit(run(sys.argv[1]))
